internal/backuprestore/backup.go

```diff
@@ -569,56 +568,73 @@ func (h *BRHandler) CleanupStaleBackups(ctx context.Context, backups []*velerov1
 }
 
-func (h *BRHandler) ensureBackupsDeleted(ctx context.Context, backups []velerov1.Backup) (bool, error) {
-	err := wait.PollUntilContextTimeout(ctx, 1*time.Second, 5*time.Minute, true,
+func (h *BRHandler) waitForDeleteBackupRequests(ctx context.Context, backups []velerov1.Backup) error {
+	return wait.PollUntilContextTimeout(ctx, 1*time.Second, 5*time.Minute, true, //nolint:wrapcheck
 		func(ctx context.Context) (bool, error) {
-			var remainingBackups []string
 			for _, backup := range backups {
-				err := h.Get(ctx, types.NamespacedName{
+				backupRequest := &velerov1.DeleteBackupRequest{}
+				if err := h.Get(ctx, types.NamespacedName{
 					Name:      backup.Name,
 					Namespace: backup.Namespace,
-				}, &velerov1.Backup{})
-				if err != nil {
+				}, backupRequest); err != nil {
 					if !k8serrors.IsNotFound(err) {
-						return false, fmt.Errorf("failed to get backup %s: %w", backup.GetName(), err)
+						return false, fmt.Errorf("failed to get deletebackuprequest %s: %w", backup.GetName(), err)
 					}
 				} else {
-					// Backup still exists
-					remainingBackups = append(remainingBackups, backup.Name)
+					if backupRequest.Status.Phase != velerov1.DeleteBackupRequestPhaseProcessed {
+						h.Log.Info("DeleteBackupRequest is not processed yet", "backup", backupRequest.GetName(), "phase", backupRequest.Status.Phase)
+						return false, nil
+					}
+					if len(backupRequest.Status.Errors) != 0 {
+						return true, fmt.Errorf("deleteBackupRequest %s failed with errors: %v", backupRequest.GetName(), backupRequest.Status.Errors)
+					}
 				}
 			}
-
-			if len(remainingBackups) == 0 {
-				h.Log.Info("All backups have been deleted")
-				return true, nil
-			}
-
-			h.Log.Info("Waiting for backups to be deleted", "backups", remainingBackups)
-			return false, nil
+			return true, nil
 		})
-	if err != nil {
+}
+
+func (h *BRHandler) ensureBackupsDeleted(ctx context.Context, backups []velerov1.Backup) error {
+	if err := h.waitForDeleteBackupRequests(ctx, backups); err != nil {
 		if errors.Is(err, context.DeadlineExceeded) {
 			h.Log.Error(err, "Timeout waiting for backups to be deleted")
-			return false, nil
+			return err
+		}
+		return fmt.Errorf("failed deleting backups: %w", err)
+	}
+
+	for _, backup := range backups {
+		err := common.RetryOnRetriable(common.RetryBackoffTwoMinutes, func() error {
+			return h.Get(ctx, types.NamespacedName{ //nolint:wrapcheck
+				Name:      backup.Name,
+				Namespace: backup.Namespace,
+			}, &velerov1.Backup{})
+		})
+		if err != nil {
+			if k8serrors.IsNotFound(err) {
+				continue
+			}
+			return fmt.Errorf("failed to ensure backup %s is deleted: %w", backup.GetName(), err)
+		} else {
+			return fmt.Errorf("all DeleteBackupRequests are processed, but backup %s is not deleted", backup.GetName())
 		}
-		return false, fmt.Errorf("api call errors when tring to ensure backup deletion: %w", err)
 	}
-	return true, nil
+	return nil
```

Comment:
Could we add a log to indicate that all backups are cleaned up?

Reply:
updated